pkg/storage/secret/keeper_store.go

```diff
@@ -153,7 +153,26 @@ func (s *keeperStorage) Delete(ctx context.Context, nn xkube.NameNamespace) erro
 	}
 
 	row := &keeperDB{Name: nn.Name, Namespace: nn.Namespace.String()}
-	err := s.db.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+
+	err := s.db.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		// Before deleting a keeper, we need to check that no secure values are referencing it.
+		secureValueCond := &secureValueDB{Keeper: row.Name, Namespace: row.Namespace}
+		secureValueRows := make([]*secureValueDB, 0)
+
+		if err := sess.Table(secureValueCond.TableName()).ForUpdate().Find(&secureValueRows, secureValueCond); err != nil {
```

Comment:
Do we need the `ForUpdate` here? Someone could insert a new secret referencing the keeper just after the select returns, right?

Reply:
Good point! I will remove it

Reply:
Good point! I guess then we would have to lock the whole `secure_value` table? 🤔 Otherwise we could have this case and then delete the keeper but references to it still exist.

Reply:
I'm assuming we don't have foreign keys because of sqlite, is that right?

I think locking the whole table may be a problem

Reply:
Technically we could but it's been discouraged in the codebase https://github.com/grafana/grafana/blob/98e9f3a534429cf98a92f05948618bfa26b6ec81/contribute/backend/style-guide.md?plain=1#L207

> I think locking the whole table may be a problem

Agreed!

Reply:
The comment is from 2018, maybe something changed since then?